### PR TITLE
Add SwerveDrive system plugin

### LINKED ISSUE #3364: Feature Request: Add SwerveDrive system plugin
## Desired behavior

Gazebo currently provides built-in drive systems such as AckermannSteeringDrive, DiffDrive and MecanumDrive, but does not include a native implementation for a swerve drive (four-wheel independent steering, 4WISD) system.

I would like to propose adding a SwerveDrive system plugin to gz-sim with the following capabilities:

- Four independently steerable wheels
- Independent wheel velocity control
- `cmd_vel` interface support
- Omnidirectional motion capability
- Consistent odometry estimation

Swerve drive systems are increasingly common in modern mobile robotics platforms due to their maneuverability and efficiency. Providing native support would expand gz-sim’s coverage of mobile robot kinematic models.

## Implementation suggestion

The proposed SwerveDrive system would:

- Follow the architectural pattern of existing drive systems (DiffDrive, MecanumDrive, AckermannDrive)
- Compute steering angles and wheel velocities from commanded twist
- Publish odometry based on wheel states
- Include integration tests and example world files

To maintain consistency with other drive systems, it may be preferable to introduce a reusable `SwerveDriveOdometry` utility in `gz-math`, similar to existing odometry helpers for other drive types.

### PR DESCRIPTION
**Pull request by C88-YQ**
# 🎉 New feature

Closes #3364 
Related to https://github.com/gazebosim/gz-math/pull/718

## Summary

This PR introduces a new SwerveDrive system plugin to gz-sim, providing support for four-wheel independent steering (swerve / 4WISD) kinematics.

The new system:

- Supports four independently steerable wheels
- Computes steering angles and wheel velocities from cmd_vel input
- Publishes odometry based on wheel states
- Follows the architectural pattern of existing drive systems such as DiffDrive and MecanumDrive
- Includes integration tests and an example world file

Kinematic and odometry computation are handled by the `SwerveDriveOdometry` utility introduced in `gz-math` (https://github.com/gazebosim/gz-math/pull/718).

This implementation does not modify existing drive systems and is fully backward compatible.

## Test it

1. Make sure the corresponding gz-math PR has been merged (or build against a local branch containing `SwerveDriveOdometry`).
2. Build gz-sim
3. Run tests:
    ``` bash
    ctest --output-on-failure
    ````
4. Manually test using the example world:
    ``` bash
    gz sim swerve_drive.sdf
    ````
5. Publish velocity commands:
    ``` bash
    gz topic -t "/model/vehicle_blue/cmd_vel" -m gz.msgs.Twist -p "linear: {y: 0.5}, angular: {z: 0.2}"
    ````

## Checklist
- [x] Signed all commits for DCO
- [x] Added a screen capture or video to the PR description that demonstrates the feature
- [x] Added tests
- [x] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [x] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] Updated Bazel files (if adding new files). Created an issue otherwise.
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers
- [ ] Was GenAI used to generate this PR? If so, make sure to add "Generated-by" to your commits. (See [this policy](https://osralliance.org/wp-content/uploads/2025/05/OSRF-Policy-on-the-Use-of-Generative-Tools-Generative-AI-in-Contributions.pdf) for more info.)

https://github.com/user-attachments/assets/1ae04fde-6ac3-4e23-99a4-0239258b72e2

https://github.com/user-attachments/assets/cdbafeb2-62bd-4769-b93f-da2e83710cd9

### CODE REVIEW

**Comment by C88-YQ** [comment]
This PR currently depends on the new SwerveDriveOdometry utility proposed in gz-math PR https://github.com/gazebosim/gz-math/pull/718

Should I temporarily disable or stub the odometry part in gz-sim until the math PR is merged to keep CI fully independent?

**Comment by arjo129** [comment]
Thank you for your contribution. It is indeed a useful one. Can we consider targetting main instead of `gz-sim10`. Ive reviewed your gz-math pr. Its a great first attempt theres so\me minor issues with angle wrapping.

**Comment by strangerwhoisharborofdoom** [comment]
Great work on implementing the SwerveDrive plugin! This looks like a valuable addition to gz-sim's drive system plugins.

A few suggestions:

1. **Documentation**: The "Updated documentation" checkbox is unchecked. Do you plan to add docs for SwerveDrive similar to DiffDrive/MecanumDrive?

2. **Testing**: Consider adding edge case tests for zero velocity commands and maximum steering angles.

3. **Example World**: Consider adding comments in the .sdf file explaining key parameters for users.

The implementation follows existing drive system patterns well. Looking forward to seeing this merged!

**Comment by C88-YQ** [comment]
Hi @arjo129 , I wanted to ask about the best way to handle this PR given its dependency on `SwerveDriveOdometry` in `gz-math`.

At the moment, this PR cannot be merged yet because CI depends on a new gz-math release. It also seems difficult to backport, since the required gz-math changes are not ready to be backported to other versions for now.

In this case, should I temporarily disable the odometry-related part in this PR, so the main SwerveDrive functionality can move forward independently? Then I could open a follow-up PR to re-enable odometry once the corresponding gz-math release is available.

**Comment by arjo129** [comment]
No it's fine we will release gz-math and things will get fixed

**Comment by arjo129** [comment]
Would we perhaps point this to main.

**Comment by C88-YQ** [comment]
Hi @arjo129 , this PR is already targeting main now ：）

**Comment by C88-YQ** [comment]
Hi @arjo129 
I noticed that the test passes on Linux and Windows CI, but fails on macOS (`gz_sim-ci-pr_any-homebrew-arm64`) for:
﻿
`SwerveDriveTest.
PublishCmd`
﻿
What confuses me is that this test is symmetric: it sends one positive velocity command and one negative velocity command. So intuitively, if there were a real bug in the command handling logic, I would expect both directions to fail similarly. But the CI output consistently points to this failure only:
﻿
```
/Users/jenkins/jenkins-agent/workspace/gz_sim-ci-pr_any-homebrew-arm64/gz-sim/test/integration/swerve_drive_system.cc:226
The difference between poses.back().Pos().X() and finalModelFramePose.Pos().X() is 0.36673264616770818, which exceeds 3e-2, where
poses.back().Pos().X() evaluates to 0.36673264566697106,
finalModelFramePose.Pos().X() evaluates to -5.0073709068065853e-10, and
3e-2 evaluates to 0.029999999999999999.
```
﻿
I do not have easy access to a macOS setup right now, so I wanted to ask if you have any suggestions on possible directions to fix this test.
﻿
I also noticed that in `DiffDrive`, similar tests are restricted with `GZ_UTILS_TEST_ENABLED_ONLY_ON_LINUX`. I looked into the related issues, but I’m not sure whether my case is same.
﻿
Would it be appropriate to handle this in a similar way, or is there a better approach here?

**Comment by C88-YQ** [comment]
Hi @arjo129
I noticed that my test passes consistently on `gz_sim-ci-pr_any-noble-amd64` and `gz_sim-pr-cnlwin`, but fails on `gz_sim-ci-pr_any-homebrew-arm64`.

After carefully checking the logs, I found that the issue only occurs in the pure lateral motion cases:
``` c++
testCmdVel(0., +1., 0.); /* pure left */
testCmdVel(0., -1., 0.); /* pure right */
```
From the logs:
``` bash
247: /Users/jenkins/jenkins-agent/workspace/gz_sim-ci-pr_any-homebrew-arm64/gz-sim/test/integration/swerve_drive_system.cc:226: Failure
247: The difference between poses.back().Pos().X() and finalModelFramePose.Pos().X() is 0.36673264616770818, which exceeds 3e-2, where
247: poses.back().Pos().X() evaluates to 0.36673264566697106,
247: finalModelFramePose.Pos().X() evaluates to -5.0073709068065853e-10, and
247: 3e-2 evaluates to 0.029999999999999999.
247: /Users/jenkins/jenkins-agent/workspace/gz_sim-ci-pr_any-homebrew-arm64/gz-sim/test/integration/swerve_drive_system.cc:227: Failure
247: The difference between poses.back().Pos().Y() and finalModelFramePose.Pos().Y() is 0.42485547384344957, which exceeds 3e-2, where
247: poses.back().Pos().Y() evaluates to 0.89870382620539591,
247: finalModelFramePose.Pos().Y() evaluates to 1.3235593000488455, and
247: 3e-2 evaluates to 0.029999999999999999.


247: /Users/jenkins/jenkins-agent/workspace/gz_sim-ci-pr_any-homebrew-arm64/gz-sim/test/integration/swerve_drive_system.cc:226: Failure
247: The difference between poses.back().Pos().X() and finalModelFramePose.Pos().X() is 0.39514500602277614, which exceeds 3e-2, where
247: poses.back().Pos().X() evaluates to 0.39514500559673849,
247: finalModelFramePose.Pos().X() evaluates to -4.2603764960347235e-10, and
247: 3e-2 evaluates to 0.029999999999999999.
247: /Users/jenkins/jenkins-agent/workspace/gz_sim-ci-pr_any-homebrew-arm64/gz-sim/test/integration/swerve_drive_system.cc:227: Failure
247: The difference between poses.back().Pos().Y() and finalModelFramePose.Pos().Y() is 0.42543242502104461, which exceeds 3e-2, where
247: poses.back().Pos().Y() evaluates to -0.89812687500634247,
247: finalModelFramePose.Pos().Y() evaluates to -1.3235593000273871, and
247: 3e-2 evaluates to 0.029999999999999999.
```

This indicates that during pure lateral motion, the odometry-based pose deviates significantly from the ground truth model pose in both X and Y directions (around 0.3–0.4 m), exceeding the current tolerance.

Since the test passes reliably on Linux (`gz_sim-ci-pr_any-noble-amd64`) and Win (`gz_sim-pr-cnlwin`), I suspect this may be related to platform-dependent behavior.

At the moment, I haven’t found a clear solution. But I noticed that in [DiffDrive](https://github.com/gazebosim/gz-sim/blob/7b12436b82828c03f5e8c9eed7c3c20472e51019/test/integration/diff_drive_system.cc#L221), similar tests are restricted using `GZ_UTILS_TEST_ENABLED_ONLY_ON_LINUX`. I looked into the related issue, but I’m still not sure whether this situation is same as mine.

So I would like to ask if there is a recommended way to handle this kind of issue. Any suggestions would be greatly appreciated.

**Comment by arjo129** [comment]
Floating point tolerances can be a challenge across architectures. I think you can simply disable the test on macOS if it is just floating point issues.

File: test/integration/swerve_drive_system.cc [hunk 1]
```diff
@@ -0,0 +1,498 @@
...
+  msgs::Set(msg.mutable_angular(), math::Vector3d(0.0, 0, 0.2));
+
+  pub.Publish(msg);
+
+  server.Run(true, 100, false);
+
+  int sleep = 0;
+  int maxSleep = 30;
+  // cppcheck-suppress knownConditionTrueFalse
+  for (; odomPosesCount < 5 && sleep < maxSleep; ++sleep)
```
Comment (arjo129) [inline test/integration/swerve_drive_system.cc]: Consider using this: https://github.com/gazebosim/gz-sim/blob/main/test/helpers/Subscription.hh

File: test/integration/swerve_drive_system.cc [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+
+          EXPECT_DOUBLE_EQ(msgTime, lastMsgTime + period);
+          lastMsgTime = msgTime;
+
+          odomPoses.push_back(msgs::Convert(_msg.pose()));
+        };
+
+        //  Publish command and check that vehicle moved
+        transport::Node node;
+      auto pub = node.Advertise<msgs::Twist>(_cmdVelTopic);
```
Comment (arjo129) [inline test/integration/swerve_drive_system.cc:107]: Wonder what happened to the style here.

File: test/integration/swerve_drive_system.cc [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+      test::Relay velocityRamp;
+      double desiredLinVelX = fb_component;
+      double desiredLinVelY = lr_component;
+      double desiredAngVel  = yaw_component;
+
+      velocityRamp.OnPreUpdate(
+          [&](const UpdateInfo &/*_info*/,
+              const EntityComponentManager &)
+          {
+            msgs::Set(msg.mutable_linear(),
```
Comment (arjo129) [inline test/integration/swerve_drive_system.cc]: Why are we publishing from within a system.

File: test/integration/swerve_drive_system.cc [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+      if (std::abs(yaw_component) > 0.0)
+      {
+        // the relative rotation has the same sign as the commanded rotation
+        EXPECT_GT(desiredAngVel * relativeMotionRot, 0.0);
+      }
+
+      // The value from odometry will be close, but not exactly the ground truth
+      // pose of the robot model. This is partially due to throttling the
+      // odometry publisher, and partially due to wheel slip.
+      EXPECT_NEAR(poses[1020].Pos().X(), odomPoses[0].Pos().X(), 3e-2);
```
Comment (arjo129) [inline test/integration/swerve_drive_system.cc:215]: This could potentially flake.

**Review by arjo129 — CHANGES_REQUESTED** [review]
The test in this PR could be improved. Probably by using this helper: https://github.com/gazebosim/gz-sim/blob/main/test/helpers/Subscription.hh

**Comment by cheriehu** [comment]
This seems to work for me! There was a little slippage (I think) at around the 15 second mark but I couldn't replicate it again. 

(changed the light source to be at the bottom to see the wheels)

https://github.com/user-attachments/assets/e92ed6cc-bfde-4160-be93-b3a35f92c300

File: src/systems/swerve_drive/SwerveDrive.cc [hunk 5]
```diff
@@ -0,0 +1,1118 @@
...
+  if (this->dataPtr->frontLeftWheelJoints.empty() ||
+      this->dataPtr->frontRightWheelJoints.empty() ||
+      this->dataPtr->backLeftWheelJoints.empty() ||
+      this->dataPtr->backRightWheelJoints.empty() ||
+      this->dataPtr->frontLeftSteeringJoints.empty() ||
+      this->dataPtr->frontRightSteeringJoints.empty() ||
+      this->dataPtr->backLeftSteeringJoints.empty() ||
+      this->dataPtr->backRightSteeringJoints.empty())
+  {
+    return;
```
Comment (arjo129) [inline src/systems/swerve_drive/SwerveDrive.cc:640]: Log an error

File: src/systems/swerve_drive/SwerveDrive.cc [hunk 6]
```diff
@@ -0,0 +1,1118 @@
...
+    gzmsg << "Found joints for model [" << modelName
+           << "], plugin will start working." << std::endl;
+    warnedModels.erase(modelName);
+  }
+
+  // Nothing left to do if paused.
+  if (_info.paused)
+    return;
+
+  for (Entity joint : this->dataPtr->frontLeftWheelJoints)
```
Comment (arjo129) [inline src/systems/swerve_drive/SwerveDrive.cc]: Im curious why we support multiple joints per wheel.

File: src/systems/swerve_drive/SwerveDrive.cc [hunk 7]
```diff
@@ -0,0 +1,1118 @@
...
+  }
+
+  for (Entity joint : this->dataPtr->backRightWheelJoints)
+  {
+    // Update wheel velocity
+    _ecm.SetComponentData<components::JointVelocityCmd>(joint,
+      {this->dataPtr->backRightWheelJointSpeed});
+  }
+
+  for (Entity joint : this->dataPtr->frontLeftSteeringJoints)
```
Comment (arjo129) [inline src/systems/swerve_drive/SwerveDrive.cc]: Same for steering: why support multiple joints. Is one not enough?

File: src/systems/swerve_drive/SwerveDrive.cc [hunk 8]
```diff
@@ -0,0 +1,1118 @@
...
+
+  for (Entity joint : this->dataPtr->backRightSteeringJoints)
+  {
+    // Update steering velocity
+    _ecm.SetComponentData<components::JointVelocityCmd>(joint,
+      {this->dataPtr->backRightSteeringJointSpeed});
+  }
+
+  // Create the joint position components if they don't exist.
+  auto frontLeftWheelPos = _ecm.Component<components::JointPosition>(
```
Comment (arjo129) [inline src/systems/swerve_drive/SwerveDrive.cc]: Consider using https://github.com/gazebosim/gz-sim/blob/19be12c1aa279bbaf6ec6ab0e58fd2d995805984/include/gz/sim/Util.hh#L273

File: src/systems/swerve_drive/SwerveDrive.cc [hunk 9]
```diff
@@ -0,0 +1,1118 @@
...
+  /// \param[in,out] _steeringDeltaAngle The steering angle change to optimize
+  /// \param[in,out] _wheelSpeed The wheel speed to optimize
+  private: void OptimizeWheelCmd(double &_steeringDeltaAngle,
+    double &_wheelSpeed);
+
+  /// \brief Gazebo communication node.
+  public: transport::Node node;
+
+  /// \brief Entity of the front left wheel joint
+  public: std::vector<Entity> frontLeftWheelJoints;
```
Comment (arjo129) [inline src/systems/swerve_drive/SwerveDrive.cc]: Why `vector` instead of `optional` ?

**Review by arjo129 — CHANGES_REQUESTED** [review]
There are several code related issues that need fixing before we can consider meging this.

**Comment by arjo129** [comment]
Thanks @cheriehu for the behaviorial test!!

**Comment by C88-YQ** [comment]
Thanks @arjo129 for the review, and thanks @cheriehu for testing it!

I’m a bit busy these days, but I’ll try to fix the remaining issues this weekend or next.

File: test/integration/swerve_drive_system.cc (hunk 1 above)
Comment (C88-YQ) [reply test/integration/swerve_drive_system.cc]: Thanks for the suggestion! Done :smile:

File: test/integration/swerve_drive_system.cc (hunk 2 above)
Comment (C88-YQ) [reply test/integration/swerve_drive_system.cc:107]: Thanks for the reminder!

File: src/systems/swerve_drive/SwerveDrive.cc (hunk 6 above)
Comment (C88-YQ) [reply src/systems/swerve_drive/SwerveDrive.cc]: Thanks for pointing this out. You’re right—supporting one joint per wheel is sufficient. I’ve updated the implementation accordingly.

File: src/systems/swerve_drive/SwerveDrive.cc (hunk 9 above)
Comment (C88-YQ) [reply src/systems/swerve_drive/SwerveDrive.cc]: @arjo129 Thanks for pointing this out. The controller currently targets a **four-wheel swerve-drive** configuration, where each wheel and steering position has exactly one joint. I’ve therefore changed the joint members from `std::vector<Entity>` to `Entity`.

File: src/systems/swerve_drive/SwerveDrive.cc (hunk 8 above)
Comment (C88-YQ) [reply src/systems/swerve_drive/SwerveDrive.cc]: Thanks for the suggestion! Done :smile:

File: src/systems/swerve_drive/SwerveDrive.cc (hunk 5 above)
Comment (C88-YQ) [reply src/systems/swerve_drive/SwerveDrive.cc:640]: @arjo129 Thanks for your suggestion.

The missing joints are already logged above with specific warnings, currently around lines 489–626, so I’m not sure whether an additional log is needed here. Did you mean adding a message such as:

```cpp
gzdbg << "Waiting for all wheel and steering joints for model ["
      << modelName << "]" << std::endl;
```

File: test/integration/swerve_drive_system.cc (hunk 4 above)
Comment (C88-YQ) [reply test/integration/swerve_drive_system.cc:215]: @arjo129 
Thanks for pointing this out. I used this approach partly because similar checks are used in the [DiffDrive](https://github.com/gazebosim/gz-sim/blob/d84b1e475577b821a5ba4a6818722a7afc769075/test/integration/diff_drive_system.cc#L193-L194), [AckermannSteering](https://github.com/gazebosim/gz-sim/blob/d84b1e475577b821a5ba4a6818722a7afc769075/test/integration/ackermann_steering_system.cc#L195-L196), and [MecanumDrive](https://github.com/gazebosim/gz-sim/blob/d84b1e475577b821a5ba4a6818722a7afc769075/test/integration/mecanum_drive_system.cc#L232-L233) tests.

Earlier in the test, we verify that running the server for 4000 iterations produces exactly 4000 recorded poses,
[L137](https://github.com/C88-YQ/gz-sim/blob/e1867a4e83a09752044fcfe5f580597329f7907a/test/integration/swerve_drive_system.cc#L137). We also check that the odometry message timestamps increase from the start time of 1 s at the configured 50 Hz rate, meaning a 0.02 s interval between messages, [L148-L164](https://github.com/C88-YQ/gz-sim/blob/e1867a4e83a09752044fcfe5f580597329f7907a/test/integration/swerve_drive_system.cc#L148-L164). Based on those checks, the correspondence between `poses[1020]` and `odomPoses[0]` should be deterministic.

Do you think there is still a case where this index could become unstable or cause the test to flake?Alternatively, I could store the recorded poses as `gz::msgs::Pose_V` with timestamps and, instead of using the fixed index 1020, find the pose whose timestamp is closest to the odometry message timestamp.

File: test/integration/swerve_drive_system.cc (hunk 3 above)
Comment (C88-YQ) [reply test/integration/swerve_drive_system.cc]: I originally followed the pattern used in some of the other drive system tests, but looking at it again, it does seem unnecessary and a bit odd. I’ve updated it to publish directly with `pub.Publish(msg)`. :smile:

**Comment by cheriehu** [comment]
Did another behavioral test and still works 🙌

**Review by arjo129 — APPROVED** [review]
Thanks for the contribution! It is a very nice one. Sorry for the long review time.

**Comment by C88-YQ** [comment]
@arjo129  No worries at all! I’m just really happy to see this PR finally merged! 🎉

Thank you all so much for the helpful suggestions, detailed reviews, and support throughout the process. I really appreciate it!
